common/include/pcl/common/common.h

```diff
@@ -262,6 +262,41 @@ namespace pcl
   PCL_EXPORTS void
   getMeanStdDev (const std::vector<float> &values, double &mean, double &stddev);
 
+  /** \brief Compute the median of a list of values (fast). If the number of values is even, take the mean of the two middle values.
+    * This function can be used like this:
+    * \code{.cpp}
+    * std::vector<double> vector{1.0, 25.0, 9.0, 4.0, 16.0};
+    * const double median = pcl::computeMedian (vector.begin (), vector.end (), static_cast<double(*)(double)>(std::sqrt)); // = 3
+    * \endcode
+    * \param[in,out] begin,end Iterators that mark the beginning and end of the value range. These values will be reordered!
+    * \param[in] f A lamda, function pointer, or similar that is implicitly applied to all values before median computation. In reality, it will be applied lazily (i.e. at most twice) and thus may not change the sorting order (e.g. monotonic functions like sqrt are allowed)
+    * \return the median
+    * \ingroup common
+    */
+  template<typename iterator, typename Functor> inline auto
+  computeMedian (iterator begin, iterator end, Functor f) noexcept -> typename std::iterator_traits<iterator>::value_type
+  {
+    const std::size_t size = std::distance(begin, end);
+    const std::size_t mid = size/2;
+    if (size%2==0)
+    { // Even number of values
+      std::nth_element (begin, begin + (mid-1), end);
+      return (f(begin[mid-1]) + f(*(std::min_element (begin + mid, end)))) / 2.0;
+    }
+    else
+    { // Odd number of values
+      std::nth_element (begin, begin + mid, end);
+      return f(begin[mid]);
+    }
+  }
+
+  /** \brief Compute the median of a list of values (fast). See the other overloaded function for more information.
+    */
+  template<typename iterator> inline auto
+  computeMedian (iterator begin, iterator end) noexcept -> typename std::iterator_traits<iterator>::value_type
+  {
+    return computeMedian (begin, end, [](const typename std::iterator_traits<iterator>::value_type& x) -> typename std::iterator_traits<iterator>::value_type { return x; });
```

Comment:
`reference` is also there in `iterator_traits`.

PS: Did the compiler trip on a simple lambda `[](const auto&x){return x;}`?